File: src/main/java/core/procedures/math/Numerator.java
package core.procedures.math;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.BigRatio;
import core.utils.Utils;

import java.math.BigDecimal;

public final class Numerator extends AFn {

  public Numerator() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{BigRatio.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "numerator";
  }

  @Override
  public Number apply1(Object arg) {
    return numerator(arg);
  }

  private Number numerator(Object o) {
    boolean isExact = Utils.isExact(o);
    Number exact;
    if (isExact) {
      exact = (Number)o;
    } else {
      exact = ToExact.toExact(o);
    }
    if (exact instanceof BigRatio) {
      if (!isExact) {
        BigDecimal result = new BigDecimal(((BigRatio) exact).getNumerator());
        return result.setScale(1, Utils.ROUNDING_MODE);
      }
      return ((BigRatio) exact).getNumerator();
    }
    return exact;
  }
}

File: src/main/java/core/procedures/math/Numerator.kt
package core.procedures.math

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Ratio
import core.scm.Type
import core.utils.Utils

class Numerator : AFn<Number?, Number>(name = "numerator", isPure = true, arity = Exactly(1),
                                       mandatoryArgsTypes = arrayOf(Type.Rational::class.java)) {

    private val toExact = ToExact()

    override operator fun invoke(arg: Number?) = numerator(arg!!)

    private fun numerator(o: Any): Number {
        val isIntegerOrRatio = o is Ratio || Utils.isInteger(o)
        val exact: Number = if (isIntegerOrRatio) (o as Number) else toExact(o as Number)
        if (exact is Ratio) {
            if (!isIntegerOrRatio) {
                return exact.numerator.toBigDecimal().setScale(1, Utils.ROUNDING_MODE)
            }
            return exact.numerator
        }
        return exact
    }
}
